### PR TITLE
check magic before parsing superblocks

### LINKED ISSUE #49: Check magic strings before proceeding with parsing
As seen in #34, we had an exception due to a parser running on invalid data. The headers ligned up so that one parser would read a string length from where another had an integer value.

For all parsers, check any applicable magic strings, and fail-fast, before proceeding.

Possibly allow ignoring the checks using some settings object.

Related: #34

### LINKED ISSUE #135: DetectFileSystems(...) fails on NTFS volumes "XFS Features not supported"
Hi @zivillian ,

It looks like the XFS V5 change has an issue when trying to read a windows (NTFS) vhd. Can you take a look?

Exception location:

            if (SbVersion >= (ushort)VersionFlags.Version5)
            {
                CompatibleFeatures = EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xD0);
                ReadOnlyCompatibleFeatures = (ReadOnlyCompatibleFeatures)EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xD4);
                IncompatibleFeatures = (IncompatibleFeatures)EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xD8);
                LogIncompatibleFeatures = EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xDC);
                Crc = EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xE0);
                SparseInodeAlignment = EndianUtilities.ToUInt32BigEndian(buffer, offset + 0xE4);
                ProjectQuotaInode = EndianUtilities.ToUInt64BigEndian(buffer, offset + 0xE8);
                Lsn = EndianUtilities.ToInt64BigEndian(buffer, offset + 0xF0);
                MetaUuid = EndianUtilities.ToGuidBigEndian(buffer, offset + 0xF8);
                if ((IncompatibleFeatures & IncompatibleFeatures.Supported)!= IncompatibleFeatures.Supported)
                    throw new NotSupportedException("XFS Features not supported");
            }

Stack:
>	DiscUtils.Xfs.dll!DiscUtils.Xfs.SuperBlock.ReadFrom(byte[] buffer, int offset) Line 423	C#
 	DiscUtils.Xfs.dll!DiscUtils.Xfs.XfsFileSystem.Detect(System.IO.Stream stream) Line 75	C#
 	DiscUtils.Xfs.dll!DiscUtils.Xfs.FileSystemFactory.Detect(System.IO.Stream stream, DiscUtils.VolumeInfo volume) Line 34	C#
 	DiscUtils.Core.dll!DiscUtils.FileSystemManager.DoDetect(System.IO.Stream stream, DiscUtils.VolumeInfo volume) Line 115	C#
 	DiscUtils.Core.dll!DiscUtils.FileSystemManager.DetectFileSystems(DiscUtils.VolumeInfo volume) Line 82	C#
 	TravistTestVhd.exe!TravistTestVhd.Program.Main(string[] args) Line 97	C#

### PR DESCRIPTION
**Pull request by zivillian**
Fixes #49 
Fixes #135

### CODE REVIEW

**Comment by LordMike** [comment]
@zivillian could you see if you can update this branch with the latest `master`?

**Comment by LordMike** [comment]
"Merge branch 'master' into magic" ... I lol'd :)

File: Library/DiscUtils.Ntfs/BiosParameterBlock.cs [hunk 1]
```diff
@@ -126,27 +126,29 @@ public void Dump(TextWriter writer, string linePrefix)
...
-            bpb.OemId = EndianUtilities.BytesToString(bytes, offset + 0x03, 8);
             bpb.BytesPerSector = EndianUtilities.ToUInt16LittleEndian(bytes, offset + 0x0B);
+            bpb.TotalSectors16 = EndianUtilities.ToUInt16LittleEndian(bytes, offset + 0x13);
+            bpb.TotalSectors32 = EndianUtilities.ToUInt32LittleEndian(bytes, offset + 0x20);
+            bpb.SignatureByte = bytes[offset + 0x26];
+            bpb.TotalSectors64 = EndianUtilities.ToInt64LittleEndian(bytes, offset + 0x28);
+            bpb.MftCluster = EndianUtilities.ToInt64LittleEndian(bytes, offset + 0x30);
+            bpb.RawMftRecordSize = bytes[offset + 0x40];
             bpb.SectorsPerCluster = bytes[offset + 0x0D];
+            if (!bpb.IsValid(long.MaxValue)) return bpb;
```
Comment (LordMike) [inline Library/DiscUtils.Ntfs/BiosParameterBlock.cs]: Is this called at other points in time?

Comment (zivillian) [reply Library/DiscUtils.Ntfs/BiosParameterBlock.cs]: This is also checked in [NtfsFileSystem.cs#L76](https://github.com/DiscUtils/DiscUtils/pull/104/files/274731cdc77a31f925ae50f2b16a1f83962ce519#diff-e4f867b61d259998b881101f291acc69R76) with the current filesystem length.

**Pull request merged**